### PR TITLE
Add movable() API point to DragBoxLayer

### LINKED ISSUE #365: Drag Box Feature Request - Draggable Boxes
From Robert Imig on the subject of Drag Boxes (PR #362):

> This is cool. Are you going to be able to drag the box around once you create the box? Or do you have to re-create it every time? Understand that might not be done now but hopefully we are thinking about it.

> One of the big use cases currently at a couple customers is to create the XDragBox you did below on chart and then slide that around to different times, and in a separate widget you see the contents of your sliding window.

### PR DESCRIPTION
**Pull request by jtlan**
Added the ability to move drag boxes by clicking and dragging on the inside of the box. The functionality is accessed via the `movable()` API point on `DragBoxLayer`:

```Typescript
/**
 * Gets whether or not the drag box is movable.
 */
movable(): boolean;
/**
 * Sets whether or not the drag box is movable.
 *
 * @param {boolean} movable
 * @return {DragBoxLayer} The calling DragBoxLayer.
 */
movable(movable: boolean): DragBoxLayer;
```

Example: http://jsfiddle.net/qr5t8754/
More complicated example: http://jsfiddle.net/657jg26g/

Close #365.

### CODE REVIEW

File: plottable.css [hunk 1]
```diff
@@ -143,6 +143,19 @@ svg.plottable {
...
+
+.plottable .drag-box-layer.movable .selection-area {
+  cursor: move; /* IE fallback */
+  cursor: -moz-grab;
+  cursor: -webkit-grab;
+  cursor: grab;
+}
+
+.plottable .drag-box-layer.movable .selection-area:active {
+  cursor: -moz-grabbing;
```
Comment (aicioara) [inline plottable.css]: No fallback here?

File: src/components/dragBoxLayer.ts [hunk 2]
```diff
@@ -64,37 +65,48 @@ export module Components {
...
+      var lastEndPoint: Point;
+
+      var DRAG_MODES = {
+        newBox: 0,
+        resize: 1,
+        move: 2
+      };
+      var mode = DRAG_MODES.newBox;
 
       this._dragInteraction.onDragStart((s: Point) => {
```
Comment (aicioara) [inline src/components/dragBoxLayer.ts]: Let's not abbreviate vars to single letter. Call it `startPoint` or at least `startP`

File: src/components/dragBoxLayer.ts [hunk 3]
```diff
@@ -106,6 +118,14 @@ export module Components {
           } else if (resizingEdges.left) {
             topLeft.x = e.x;
           }
+        } else {
```
Comment (aicioara) [inline src/components/dragBoxLayer.ts]: Can we do a switch case? It's easier to figure out what this else corresponds to

File: src/components/dragBoxLayer.ts [hunk 4]
```diff
@@ -64,37 +65,48 @@ export module Components {
...
-        var bounds = this.bounds();
+        bounds = this.bounds();
         // copy points so changes to topLeft and bottomRight don't mutate bounds
         topLeft = { x: bounds.topLeft.x, y: bounds.topLeft.y };
         bottomRight = { x: bounds.bottomRight.x, y: bounds.bottomRight.y };
+        lastEndPoint = s;
         this._dragStartCallbacks.callCallbacks(bounds);
       });
 
       this._dragInteraction.onDrag((s: Point, e: Point) => {
```
Comment (aicioara) [inline src/components/dragBoxLayer.ts]: Again, single letter variables

File: test/components/dragBoxLayerTests.ts [hunk 5]
```diff
@@ -475,5 +475,138 @@ describe("Interactive Components", () => {
...
+        });
+        dbl.boxVisible(true);
+        initialBounds = dbl.bounds();
+      });
+
+      it("get and set movable()", () => {
+        assert.isFalse(dbl.movable(), "defaults to false");
+        assert.strictEqual(dbl.movable(true), dbl, "setter mode returns DragBoxLayer");
+        assert.isTrue(dbl.movable(), "set to true");
+        assert.isTrue(dbl.hasClass("movable"), "\"movable\" CSS class is applied");
```
Comment (aicioara) [inline test/components/dragBoxLayerTests.ts]: Test that it did not have it before.

File: test/components/dragBoxLayerTests.ts [hunk 6]
```diff
@@ -475,5 +475,138 @@ describe("Interactive Components", () => {
...
+        // assert.isTrue(dbl.hasClass("x-movable"), "\"x-movable\" CSS class is applied");
+        // assert.isTrue(dbl.hasClass("y-movable"), "\"y-movable\" CSS class is applied");
```
Comment (aicioara) [inline test/components/dragBoxLayerTests.ts]: I guess you wanted them deleted.

File: test/components/dragBoxLayerTests.ts [hunk 7]
```diff
@@ -475,5 +475,138 @@ describe("Interactive Components", () => {
...
+        );
+        var bounds = dbl.bounds();
+        assert.strictEqual(bounds.topLeft.y, initialBounds.topLeft.y, "top edge was not moved");
+        assert.strictEqual(bounds.bottomRight.y, initialBounds.bottomRight.y, "bottom edge was not moved");
+        assert.strictEqual(bounds.topLeft.x, initialBounds.topLeft.x, "left edge was not moved");
+        assert.strictEqual(bounds.bottomRight.x, SVG_WIDTH, "right edge was repositioned");
+        svg.remove();
+      });
+
+      it("doesn't dismiss on no-op move", () => {
```
Comment (aicioara) [inline test/components/dragBoxLayerTests.ts]: Add test for it dismisses box when clicked outside of it no-op (basically same test but x: 0, y: 0 for both start drag and end drag)

File: test/components/xDragBoxLayerTests.ts [hunk 8]
```diff
@@ -105,5 +105,43 @@ describe("Interactive Components", () => {
...
+      };
+
+      dbl.bounds({
+        topLeft: topLeft,
+        bottomRight: bottomRight
+      });
+
+      var boundsBefore = dbl.bounds();
+      var dragDistance = 10;
+      var target = dbl.background();
```
Comment (aicioara) [inline test/components/xDragBoxLayerTests.ts]: Maybe inline?

File: test/testMethods.ts [hunk 9]
```diff
@@ -157,6 +157,21 @@ module TestMethods {
 
   export function triggerFakeDragSequence(target: d3.Selection<void>, start: Plottable.Point, end: Plottable.Point) {
     triggerFakeMouseEvent("mousedown", target, start.x, start.y);
+    var numSteps = 10;
+    // var steps: Plottable.Point[] = [];
+    for (var i = 0; i < numSteps; i++) {
```
Comment (aicioara) [inline test/testMethods.ts]: Start from 1?

File: test/testMethods.ts [hunk 10]
```diff
@@ -157,6 +157,21 @@ module TestMethods {
...
+      // steps.push({
+      //   x: start.x + (end.x - start.x) * i / numSteps,
+      //   y: start.y + (end.y - start.y) * i / numSteps
+      // });
```
Comment (aicioara) [inline test/testMethods.ts]: Remove commented out code

File: plottable.css (hunk 1 above)
Comment (jtlan) [inline plottable.css]: It falls back to the non-active state.

File: src/components/dragBoxLayer.ts (hunk 3 above)
Comment (jtlan) [inline src/components/dragBoxLayer.ts]: I thought about that. Sounds good.

File: test/components/dragBoxLayerTests.ts (hunk 6 above)
Comment (jtlan) [inline test/components/dragBoxLayerTests.ts]: My bad, good catch.

File: test/components/dragBoxLayerTests.ts (hunk 7 above)
Comment (jtlan) [inline test/components/dragBoxLayerTests.ts]: Added.

**Comment by aicioara** [comment]
Looks good.

**Comment by crmorford** [comment]
This looks really good!
Investigating safari further (waiting for the ipad to charge), but otherwise this looks good on Chrome/FF/IE.

**Comment by crmorford** [comment]
Looks good on mobile safari. Desktop safari isn't good, but we knew that interactions had trouble on desktop safari previously.

**Comment by jtlan** [comment]
Desktop Safari issue here: #2454.

**Pull request merged**